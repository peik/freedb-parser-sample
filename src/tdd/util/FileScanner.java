package tdd.util;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FilenameFilter;
import java.util.Arrays;
import java.util.List;

/**
 * FileScanner lists all files in a given directory.
 */
public class FileScanner {

	/** Main directory containing the files to list */
	private final String dir;

	public FileScanner(String dir) {
		this.dir = dir;
	}

	public List<String> listTxtFiles() throws FileNotFoundException {
		File dir = new File(this.dir);

		String[] txtFiles = dir.list(this.txtFilter);
		if (txtFiles == null) {
			throw new FileNotFoundException("Directory " + dir + " not found");
		}

		return Arrays.asList(txtFiles);
	}

	// Anonymous FilenameFilter class to find only .txt files
	private final FilenameFilter txtFilter = new FilenameFilter() {
		public boolean accept(File dir, String name) {
			return name.endsWith(".txt");
		}
	};

}
