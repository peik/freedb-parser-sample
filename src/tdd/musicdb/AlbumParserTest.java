package tdd.musicdb;

import java.util.List;

import org.approvaltests.Approvals;
import org.junit.Test;

import com.thoughtworks.xstream.XStream;

public class AlbumParserTest {

	@Test
	public void testCalculateSongLengths() throws Exception {
		// Data copied from 12025d03.txt
		AlbumParser parser = new AlbumParser();
		parser.offsets.add(150);
		parser.offsets.add(12143);
		parser.offsets.add(31682);
		parser.discLengthSecs = 607;
		List<Integer> lengthsSecs = parser.calculateSongLengths();
		Approvals.approve(lengthsSecs.toString());
		// Expected result:
		// 1. I'm Your Puppet 2:39 = 159s
		// 2. Me And Jacky 4:20 = 260s
		// 3. If You Don't Like Country 3:05 = 185s
		// Total time: 10:04
	}

	@Test
	public void testParse() throws Exception {
		AlbumParser parser = new AlbumParser("c00ccd0d.txt");
		Approvals.approve(new XStream().toXML(parser));
	}

	@Test
	public void testParseToListing() throws Exception {
		AlbumParser parser = new AlbumParser("c00ccd0d.txt");
		Album album = parser.getAlbum();
		Approvals.approve(album.getListing());
	}

}
