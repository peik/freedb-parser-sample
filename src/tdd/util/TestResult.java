package tdd.util;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import org.approvaltests.Approvals;

import com.thoughtworks.xstream.XStream;
import com.thoughtworks.xstream.io.xml.XppDomDriver;

/**
 * A test helper class that collects the output of a test to be able to print it
 * in text form.<br>
 * Intended to collect data to be used with the {@link Approvals} tests library
 * (http://approvaltests.sourceforge.net/).
 * 
 * @author HiQ Finland Oy (9518174)
 * @version $Id: TestResult.java 8543 2011-12-16 16:23:58Z pas $
 */
public class TestResult implements java.io.Serializable {

	private static final long serialVersionUID = 1L;
	private final List<String> output = new ArrayList<String>();
	private boolean logToConsole = false;

	/**
	 * Constructor
	 */
	public TestResult() {
	}

	/**
	 * Constructor with header text.
	 */
	public TestResult(String header) {
		log(header);
	}

	/**
	 * By default, the result is only logged to the output list, from which it
	 * can be retrieved at the end of the test using {@link #getOutputText()}.
	 * For debugging purposes, you might want to enable logging also to the
	 * console, which prints out the output to System.out as soon as it is
	 * created, not just at the end of the test.
	 */
	public void enableLogToConsole(boolean logAlsoToConsole) {
		this.logToConsole = logAlsoToConsole;
	}

	/**
	 * Log a new message to the result.
	 */
	public void log(String message) {
		if (logToConsole) {
			System.out.println(message);
		}
		output.add(message);
	}

	/**
	 * Log a new message to the result, formatting the arguments using
	 * {@link String#format(String, Object...)}.
	 */
	public void log(String formatMessage, Object... args) {
		log(String.format(formatMessage, args));
	}

	/**
	 * Log a collection of objects using their <code>#toString()</code> values.
	 */
	public void logCollection(Collection<?> objects) {
		for (Object object : objects) {
			log(String.valueOf(object));
		}
	}

	/**
	 * Append a part of a message to the end of the previous logged message. No
	 * spaces or padding is added, so this is up to the responsibility of the
	 * caller.
	 * <p>
	 * This method is useful when you want to build the message conditionally,
	 * and append more infor to the message only under certain condition, for
	 * instance after checking that some object to write to the output is not
	 * null.
	 */
	public void append(String messagePart) {
		if (logToConsole) {
			System.out.println("... " + messagePart);
		}
		int lastPos = output.size() - 1;
		String lastLoggedMessage = output.get(lastPos);
		output.set(lastPos, lastLoggedMessage + messagePart);
	}

	/**
	 * Append a part of a message to the end of the previous logged message,
	 * formatting the arguments using {@link String#format(String, Object...)}.
	 * 
	 * @see #append(String)
	 */
	public void append(String formatMessagePart, Object... args) {
		append(String.format(formatMessagePart, args));
	}

	public void logXML(Object object) {
		XStream xs = new XStream(new XppDomDriver());
		String xml = xs.toXML(object);
		log(xml);
	}

	public void logXMLCollection(Collection<?> objects) {
		XStream xs = new XStream(new XppDomDriver());
		for (Object object : objects) {
			String xml = xs.toXML(object);
			log(xml);
		}
	}

	/**
	 * Add the output of the given result to this one, joining the results
	 * together.
	 */
	public void addOutput(TestResult result) {
		this.output.addAll(result.output);
	}

	public List<String> getOutput() {
		return output;
	}

	public String getOutputText() {
		StringBuilder sb = new StringBuilder();
		for (String line : output) {
			sb.append(line).append("\n");
		}
		return sb.toString();
	}

	/**
	 * Shortcut for <code>Approvals.approve(getOutputText());</code> which
	 * should be called at the end of an {@link Approvals} test.
	 */
	public void approve() throws Exception {
		Approvals.approve(getOutputText());
	}

}
