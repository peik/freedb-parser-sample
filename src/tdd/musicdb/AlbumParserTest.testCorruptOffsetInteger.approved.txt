Adele / 21	Details
Tracks: 12
Total time: 54:33
Year: 2011
Disc-ID: Pop / c00ccd0d

 
1.	Rolling In The Deep	7:32
2.	Rumour Has It	4:10
3.	Turning Tables	4:03
4.	Don't You Remember	4:01
5.	Set Fire To The Rain	4:37
6.	He Won't Go	3:48
7.	Take It All	4:01
8.	I'll Be Waiting	5:48
9.	One And Only	5:16
10.	Lovesong	4:43
11.	Someone Like You	3:08
12.	If It Hadn't Been For Love	3:26
